### ISSUE TITLE
etcd-cpp-apiv3 doesn't work with the newest Abseil (which gRPC is going to require soon)

### ISSUE THREAD

**Issue by devjgm** [etcd-cpp-apiv3/etcd-cpp-apiv3#194]
@coryan was trying to upgrade Abseil in vcpkg in https://github.com/microsoft/vcpkg/pull/29692, but he kept running into issues where etcd was breaking with that change. I think etcd only uses Abseil indirectly because gRPC does, however, gRPC will soon require this version of Abseil as well. The newest version of Abseil requires >= C++14, which seems to be at least part of the issue.

Can someone on this project take a look at the concerns raised in https://github.com/microsoft/vcpkg/pull/29692#issuecomment-1435154537 to see if some fixes could be made to etcd in order to unblock the Abseil (and Protobuf and gRPC) upgrades in vcpkg?

**Comment by sighingnow** [comment]
Hey @devjgm @coryan,

Thanks for the effort to make etcd-app-apiv3 better. Adding a cmake option `-DCMAKE_CXX_STANDARD=14` [here](https://github.com/microsoft/vcpkg/blob/master/ports/etcd-cpp-apiv3/portfile.cmake#L12) could resolve the problem with newer abseil.

@coryan could you please take a try to include the option in [your abseil PR](https://github.com/microsoft/vcpkg/pull/29692)? Thanks!

**Comment by coryan** [comment]
Thanks for taking the time to answer.

> Adding a cmake option -DCMAKE_CXX_STANDARD=17 [here](https://github.com/microsoft/vcpkg/blob/master/ports/etcd-cpp-apiv3/portfile.cmake#L12) could resolve the problem with newer abseil.

If Abseil was compiled with C++17 that would not work.  Under `vcpkg`, Abseil is compiled with C++17 if (a) the compiler defaults to C++17 (as recent GCC versions do), or (b) the Abseil vcpkg package was compiled with the `cxx17` feature (i.e., if you say `vcpkg install abseil[cxx17]`).

I think a more promising approach would be to use `target_link_libraries()` when compiling the `etcd-cpp-apiv3` object files. Which I implemented in [this commit](https://github.com/microsoft/vcpkg/pull/29692/commits/9746d162e67c83e7a3a9024f9a2bda99e651e018).  With that change `etcd-cpp-apiv3` inherits any compiler options required by the libraries it depends on, including any `-std=c++XX`.  The only reason that patch works is because I also apply [this other patch](https://github.com/microsoft/vcpkg/pull/29692/commits/4ca560a2373dc1396dcffcb3b4ea25374e0ff806) to take advantage for `target_compile_features(.... cxx_std_${...})`.  That is only available with CMake >= 3.8, which is a safe assumption under `vcpkg`, while I noticed your project only requires CMake >= 3.3.

In short, upstreaming these changes may be complicated if you want to continue supporting CMake < 3.8.  I will leave that decision to you.

**Comment by sighingnow** [comment]
`target_link_libraries(etcd-cpp-apiv3-core-objects ....)` is a valid fix that I would backport to this repo.

> The only reason that patch works is because I also apply [this other patch](https://github.com/microsoft/vcpkg/pull/29692/commits/4ca560a2373dc1396dcffcb3b4ea25374e0ff806) to take advantage for target_compile_features(.... cxx_std_${...}).

I'm wondering if there's any reason to prefer `target_compile_features` rather than `CXX_STANDARD` property? (not very familiar with cmake)

**Comment by sighingnow** [comment]
> `target_link_libraries(etcd-cpp-apiv3-core-objects ....)` is a valid fix that I would backport to this repo.
> 
> > The only reason that patch works is because I also apply [this other patch](https://github.com/microsoft/vcpkg/pull/29692/commits/4ca560a2373dc1396dcffcb3b4ea25374e0ff806) to take advantage for target_compile_features(.... cxx_std_${...}).
> 
> I'm wondering if there's any reason to prefer `target_compile_features` rather than `CXX_STANDARD` property? (not very familiar with cmake)

Both addressed in #195. Thanks @coryan!

**Comment by coryan** [comment]
Thank you for upstreaming the changes.

> I'm wondering if there's any reason to prefer `target_compile_features` rather than `CXX_STANDARD` property? (not very familiar with cmake)

I think (and I am not sure this is a widely shared belief) that `target_compile_features()` better captures the requirements. 
 These are often "My library requires with C++ >= XX", where are `CXX_STANDARD` says "change the compiler to use C++ YY *in this build*. 

For example, say some library requires C++ >= 14.  Using `target_compile_features(my-lib PUBLIC cxx_std_14)` will add the necessary `-std=c++14` option if the compiler defaults to C++98 or C++11, but will leave the compiler command-line as-is if the compiler defaults to C++14 or C++17.  On platforms where the default is 17, your library is more likely to work well with the rest of the libraries in that system.

Furthermore, any programs that consume your library (directly or indirectly) will also inherit the same `targt_compile_features()`. That means it is less likely that the consumers of your C++YY library will use the incorrect `CXX_STANDARD` for their program.

It is not perfect.  Libraries that change their ABI based on the C++ standard will break.  But they break with `CXX_STANDARD` too.

**Comment by sighingnow** [comment]
> Say some library requires C++ >= 14. Using target_compile_features(my-lib PUBLIC cxx_std_14) will add the necessary -std=c++14 option if the compiler defaults to C++98 or C++11, but will leave the compiler command-line as-is if the compiler defaults to C++14 or C++17. On platforms where the default is 17, your library is more likely to work well with the rest of the libraries in that system. CXX_STANDARD=14 will always build with C++14, even if the default for that compiler is C++17.

Very clear to me. Thanks!

I have added the branch for `cmake>=3.8` to use `target_compile_features` in #195.

**Comment by devjgm** [comment]
Thanks!
